longruntests/longrun_test.go

```diff
@@ -0,0 +1,200 @@
+package longruntests_test
+
+import (
+	"os"
+	"time"
+
+	. "github.com/deis/workflow-e2e/utils"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gbytes"
+	. "github.com/onsi/gomega/gexec"
+)
+
+func pingApp(pingChan chan<- bool, testApp App) {
+	defer GinkgoRecover()
+	defer func() { pingChan <- true }()
+	startTime := time.Now()
+	duration := time.Duration(1) * time.Minute
+	for time.Since(startTime) <= duration {
+		sess, err := Run(`curl -sL -w "%%{http_code}\\n" "%s" -o /dev/null`, testApp.URL)
+		Expect(err).NotTo(HaveOccurred())
+		Eventually(sess).Should(Say("200"))
+		Eventually(sess).Should(Exit(0))
+	}
+}
+
+func scaleApp(scaleChan chan<- bool, testApp App) {
+	scaleTo := 5
+	defer GinkgoRecover()
+	defer func() { scaleChan <- true }()
+	startTime := time.Now()
+	duration := time.Duration(1) * time.Minute
+	for time.Since(startTime) <= duration {
+		sess, err := Run("deis ps:scale web=%d --app=%s", scaleTo, testApp.Name)
+		Expect(err).NotTo(HaveOccurred())
+		Eventually(sess).Should(Say("Scaling processes... but first,"))
+		Eventually(sess, DefaultMaxTimeout).Should(Say(`done in \d+s`))
+		Eventually(sess).Should(Say("=== %s Processes", testApp.Name))
+		Eventually(sess).Should(Exit(0))
+
+		// test that there are the right number of processes listed
+		procsListing := ListProcs(testApp).Out.Contents()
+		procs := ScrapeProcs(testApp.Name, procsListing)
+		Expect(len(procs)).To(Equal(scaleTo))
+		if scaleTo == 5 {
+			scaleTo = 1
+		} else {
+			scaleTo = 5
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+var _ = Describe("Longrun", func() {
+
+	It("Run example Go app", func() {
+		var testApp App
+		os.Chdir("example-go")
+		for {
```

Comment:
Duplication could be vastly reduced here.  One approach would to be set these up in a table-driven style; something like [this](https://gist.github.com/vdice/213ebdfda93289e3e9518557cf3a3b09)